### PR TITLE
modules/py_ml: Add support for multi-input models.

### PR DESCRIPTION
**Pull request by iabdalkader**

### CODE REVIEW

File: scripts/libraries/ml/ml/preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        if len(args) == 1:
+            img = args[0]
+            if not isinstance(img, image.Image):
+                raise ValueError("Expected an image input")
+            if self.roi is None:
+                self.roi = (0, 0, img.width(), img.height())
+            return Normalization(img, self.scale, self.mean, self.stdev, self.roi)
+        buffer, shape, dtype = args
+        # Create an image using the input tensor as buffer.
+        img = image.Image(shape[1], shape[2], self.image.format(), buffer=buffer)
```
Comment (kwagyeman) [inline scripts/libraries/ml/ml/preprocessing.py:36]: shape[1] is h, shape[2] is w. These should be swapped.

File: src/omv/modules/py_ml.c [hunk 2]
```diff
@@ -63,175 +37,67 @@ static size_t py_ml_tuple_sum(mp_obj_tuple_t *o) {
...
+            if (model->input_dtype == 'f') {
+                float *model_input_float = (float *) input_buffer;
+                for (size_t i = 0; i < input_array->len; i++) {
+                    float value = ndarray_get_float_index(input_array->array, input_array->dtype, i);
+                    model_input_float[i] = value;
                 }
-                #endif
-                for (; size >= 0; size -= 1) {
-                    model_input_8[size] ^= PY_ML_GRAYSCALE_MID;
+            } else if (model->input_dtype == 'b') {
```
Comment (kwagyeman) [inline src/omv/modules/py_ml.c:80]: You can merge 'b' and 'B' into one method.

File: src/omv/modules/py_ml.c [hunk 3]
```diff
@@ -274,21 +140,19 @@ static const mp_obj_type_t py_ml_model_type;
...
+              self->size, self->memory_size, self->inputs_size, self->input_dtype,
+              (double) self->input_scale, self->input_zero_point, self->outputs_size, self->output_dtype,
               (double) self->output_scale, self->output_zero_point);
 }
 
 static mp_obj_t py_ml_model_predict(uint n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
-    enum { ARG_roi, ARG_callback, ARG_scale, ARG_mean, ARG_stdev };
+    enum { ARG_roi, ARG_callback };
     static const mp_arg_t allowed_args[] = {
         { MP_QSTR_roi, MP_ARG_OBJ | MP_ARG_KW_ONLY, {.u_rom_obj = MP_ROM_NONE} },
```
Comment (kwagyeman) [inline src/omv/modules/py_ml.c:154]: Remove roi.

**Review by kwagyeman — COMMENTED** [review]
Looks good. Is the callback stuff needed anymore? I removed it in my PR. You can just have the code directly in predict().

**Comment by iabdalkader** [comment]
> Looks good. Is the callback stuff needed anymore? I removed it in my PR. You can just have the code directly in predict().

No the callbacks are not needed anymore. I was going to remove them and do some clean-ups in a separate PR, but I will do this in this PR.

File: scripts/libraries/ml/ml/preprocessing.py (hunk 1 above)
Comment (iabdalkader) [reply scripts/libraries/ml/ml/preprocessing.py:36]: Good catch!

File: src/omv/modules/py_ml.c (hunk 2 above)
Comment (iabdalkader) [reply src/omv/modules/py_ml.c:79]: I prefer to leave this code as is.

File: src/omv/modules/py_ml.c (hunk 3 above)
Comment (iabdalkader) [reply src/omv/modules/py_ml.c]: Removed.

**Pull request merged**